src/cache.js

```diff
@@ -26,38 +26,61 @@ class Cache {
     this.models[modelName] = modelSchema(tableName)
   }
 
-  updateUserWithLoan (userID, loanID) {
-    return this.updateUserWithItem(userID, loanID, 'loan')
-  }
-
-  updateUserWithRequest (userID, requestID) {
-    return this.updateUserWithItem(userID, requestID, 'request')
-  }
-
-  updateUserWithItem (userID, itemID, itemType) {
+  updateUserItem (userID, itemID, operation, itemType) {
     return this.models.UserModel.get(userID)
       .then(user => {
         if (user) {
-          switch (itemType) {
-            case 'loan':
-              return user.addLoan(itemID).save()
-            case 'request':
-              return user.addRequest(itemID).save()
-          }
+          return {
+            'add': {
+              'loan': user.addLoan,
+              'request': user.addRequest
+            },
+            'delete': {
+              'loan': user.deleteLoan,
+              'request': user.deleteRequest
+            }
+          }[operation][itemType].call(user, itemID).save()
         } else {
           return this.usersQueue.sendMessage(userID)
         }
       })
   }
 
+  updateUserWithAddLoan (userID, loanID) {
+    return this.updateUserItem(userID, loanID, 'add', 'loan')
+  }
+
+  updateUserWithAddRequest (userID, requestID) {
+    return this.updateUserItem(userID, requestID, 'add', 'request')
+  }
+
+  updateUserWithDeleteLoan (userID, loanID) {
+    return this.updateUserItem(userID, loanID, 'delete', 'loan')
+  }
+
+  updateUserWithDeleteRequest (userID, requestID) {
+    return this.updateUserItem(userID, requestID, 'delete', 'request')
+  }
+
   updateLoan (itemLoan) {
     return new this.models.LoanModel(itemLoan).save()
   }
 
+  deleteLoan (loanID) {
+    return this.models.LoanModel.delete(loanID)
+  }
+
+  handleLoanReturned (itemLoan) {
+    return Promise.all([
+      this.deleteLoan(itemLoan.loan_id),
+      this.updateUserWithDeleteLoan(itemLoan.user_id, itemLoan.loan_id)
+    ])
+  }
+
   handleLoanUpdate (itemLoan) {
     return Promise.all([
       this.updateLoan(itemLoan),
-      this.updateUserWithLoan(itemLoan.user_id, itemLoan.loan_id)
+      this.updateUserWithAddLoan(itemLoan.user_id, itemLoan.loan_id)
```

Comment:
Not sure this new naming of the method is clear or not. Maybe `updateUserAndAddLoan`, not overly happy with my name suggestion.

Reply:
Possibly something along the lines of `addLoanToUser`, `deleteLoanFromUser` would be better?

Reply:
That naming is a lot clearer